javalib/src/main/scala/java/lang/String.scala

```diff
@@ -1,42 +1,951 @@
-  def concat(str: _String): _String = ???
-  def indexOf(ch: scala.Int): scala.Int = ???
-  def indexOf(ch: scala.Int, fromIndex: scala.Int): scala.Int = ???
-  def lastIndexOf(ch: scala.Int): scala.Int = ???
-  def lastIndexOf(ch: scala.Int, fromIndex: scala.Int): scala.Int = ???
-  def substring(beginIndex: scala.Int): _String = ???
-  def substring(beginIndex: scala.Int, endIndex: scala.Int): _String = ???
-  def startsWith(prefix: _String): scala.Boolean = ???
-  def toLowerCase(): _String = ???
-  def toCharArray(): Array[Char] = ???
-  def equalsIgnoreCase(str: _String): scala.Boolean = ???
-  def getChars(srcBegin: scala.Int, srcEnd: scala.Int, dst: Array[scala.Char],
-      dstBegin: scala.Int): Unit = ???
-  override def hashCode(): scala.Int = ???
-
-  // TODO: rest of the api
+class _String extends Serializable with Comparable[_String] with CharSequence {
+  private var value = new Array[Char](0)
+  private var offset = 0
+  private var count = 0
+  private override var hashCode: Int = _
+
+  private def this(s: _String, ch: Char) {
+    this()
+    offset = 0
+    value = Array.ofDim[Char](s.count + 1)
+    count = s.count + 1
+    System.arraycopy(s.value, s.offset, value, 0, s.count)
+    value(s.count) = ch
+  }
+
+  def this(data: Array[scala.Byte], start: Int, length: Int) {
+    this()
+    if (start >= 0 && 0 <= length && length <= data.length - start) {
+      offset = 0
+      val charset = defaultCharset()
+      val cb = charset.decode(ByteBuffer.wrap(data, start, length))
+      val result: Int = cb.length
+
+      if (result > 0) {
+        value = cb.array()
+        count = result
+      } else {
+        count = 0
+        value = Array.ofDim[Char](0)
+      }
+    } else {
+      throw new StringIndexOutOfBoundsException()
+    }
+  }
+
+  @Deprecated
+  def this(data: Array[scala.Byte],
+           high: Int,
+           start: Int,
+           length: Int) {
+    this()
+    if (length <= data.length - start && start >= 0 && 0 <= length) {
+      offset = 0
+      value = Array.ofDim[Char](length)
+      count = length
+      high <<= 8
+      for (i <- 0 until count) {
+        value(i) = (high + (data(start += 1) & 0xff)).toChar
+      }
+    } else {
+      throw new StringIndexOutOfBoundsException()
+    }
+  }
+
+  def this(data: Array[scala.Byte]) {
+    this(data, 0, data.length)
+  }
+
+  @Deprecated
+  def this(data: Array[scala.Byte], high: Int) {
+    this(data, high, 0, data.length)
+  }
+
+  def this(data: Array[scala.Byte],
+           start: Int,
+           length: Int,
+           encoding: _String) {
+    this()
+    if (encoding == null) {
+      throw new NullPointerException()
+    }
+
+    if (start >= 0 && 0 <= length && length <= data.length - start) {
+      offset = 0
+      val charset = getCharset(encoding)
+      val cb = Try(
+        charset.decode(ByteBuffer.wrap(data, start, length))
+      ).getOrElse(CharBuffer.wrap("?".toCharArray))
```

Comment:
Please avoid higher-level Scala constructs like `Try` for now as they are not guaranteed to be optimised away at the moment.